pkg/backend/filestate/backend.go

```diff
@@ -402,27 +491,34 @@ func (b *localBackend) Upgrade(ctx context.Context) error {
 	return nil
 }
 
-// upgradeStack upgrades a single stack to use the provided projectReferenceStore.
-func (b *localBackend) upgradeStack(
-	ctx context.Context,
-	newStore *projectReferenceStore,
-	old *localBackendReference,
-) error {
+// guessProject inspects the checkpoint for the given stack and attempts to
+// guess the project name for it.
+// Returns an empty string if the project name cannot be determined.
+func (b *localBackend) guessProject(ctx context.Context, old *localBackendReference) (tokens.Name, error) {
 	contract.Requiref(old.project == "", "old.project", "must be empty")
 
 	chk, err := b.getCheckpoint(ctx, old)
 	if err != nil {
-		return err
+		return "", fmt.Errorf("read checkpoint: %w", err)
 	}
 
 	// Try and find the project name from _any_ resource URN
-	var project tokens.Name
 	if chk.Latest != nil {
 		for _, res := range chk.Latest.Resources {
-			project = tokens.Name(res.URN.Project())
-			break
+			return tokens.Name(res.URN.Project()), nil
 		}
 	}
+	return "", nil
+}
+
+// upgradeStack upgrades a single stack to use the provided projectReferenceStore.
+func (b *localBackend) upgradeStack(
+	ctx context.Context,
+	newStore *projectReferenceStore,
+	project tokens.Name,
+	old *localBackendReference,
+) error {
+	contract.Requiref(old.project == "", "old.project", "must be empty")
 	if project == "" {
```

Comment:
Should we just make this a contract.Requires now? Make it invalid to call upgradeStack with project=""